### PR TITLE
Origin/whatsapp

### PR DESCRIPTION
**Pull request by soyombo-baterdene**

### CODE REVIEW

File: src/whatsapp/api.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  receiverId: string,
+  body: string,
+  filename: string,
+  caption: string,
+  instanceId: string,
+  token: string,
+): Promise<IMessage> => {
+  return new Promise((resolve, reject) => {
+    const requestOptions = {
+      url: `https://api.chat-api.com/instance${instanceId}/sendFile?token=${token}`,
```
Comment (munkhorgil) [inline src/whatsapp/api.ts]: https://api.chat-api.com should be constant

File: src/whatsapp/controller.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import * as whatsappUtils from './api';
+import { ConversationMessages, Conversations } from './models';
+import receiveMessage from './receiveMessage';
+
+const init = async app => {
+  app.post('/whatsapp/webhook', async (req, res, next) => {
+    try {
+      await receiveMessage(req.body);
+    } catch (e) {
+      return next(new Error(e));
```
Comment (batamar) [inline src/whatsapp/controller.ts]: remove new Error

File: src/whatsapp/controller.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    const { integrationId, data } = req.body;
+    const { instanceId, token } = JSON.parse(data);
+
+    // Check existing Integration
+
+    let integration = await Integrations.findOne({
+      $and: [{ whatsappinstanceId: instanceId }, { kind: 'whatsapp' }],
+    });
+    if (integration) {
+      return next(`Integration already exists with this instance id: ${instanceId}`);
```
Comment (batamar) [inline src/whatsapp/controller.ts]: throw new Error

File: src/whatsapp/receiveMessage.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  const instanceId = requestBody.instanceId;
+  const integration = await Integrations.getIntegration({
+    $and: [{ whatsappinstanceId: instanceId }, { kind: 'whatsapp' }],
+  });
+
+  if (requestBody.ack) {
+    for (const ack of requestBody.ack) {
+      await ConversationMessages.updateOne({ mid: ack.id }, { $set: { status: ack.status } });
+    }
+  } else if (requestBody.messages) {
```
Comment (batamar) [inline src/whatsapp/receiveMessage.ts]: use return to reduce indentation

File: src/whatsapp/receiveMessage.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+              content: message.body,
+            }),
+          });
+
+          conversation.erxesApiId = apiConversationResponse._id;
+
+          await conversation.save();
+        } catch (e) {
+          await Conversations.deleteOne({ _id: conversation._id });
+          throw new Error(e);
```
Comment (batamar) [inline src/whatsapp/receiveMessage.ts]: remove new Error

File: src/whatsapp/store.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        phones: [phoneNumber],
+        primaryPhone: phoneNumber,
+        isUser: true,
+      }),
+    });
+    customer.erxesApiId = apiCustomerResponse._id;
+    await customer.save();
+  } catch (e) {
+    await Customers.deleteOne({ _id: customer._id });
+    throw new Error(e);
```
Comment (batamar) [inline src/whatsapp/store.ts]: remove new Error

File: src/whatsapp/controller.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    return res.json({ status: 'ok' });
+  });
+
+  app.post('/whatsapp/reply', async (req, res) => {
+    const { attachments, conversationId, content, integrationId } = req.body;
+
+    if (attachments.length > 1) {
+      throw new Error('You can only attach one file');
```
Comment (batamar) [inline src/whatsapp/controller.ts]: return next

File: src/whatsapp/controller.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    const conversation = await Conversations.getConversation({ erxesApiId: conversationId });
+
+    const integration = await Integrations.findOne({ erxesApiId: integrationId });
+
+    const recipientId = conversation.recipientId;
+    const instanceId = conversation.instanceId;
+    const token = integration.whatsappToken;
+
+    if (attachments.length !== 0) {
```
Comment (batamar) [inline src/whatsapp/controller.ts]: use return

**Pull request closed without merging**